### PR TITLE
Run samples against Docker net-tools container

### PR DESCRIPTION
**Pull request by pfl**
Provide a script that runs the sample and starts a 'net-tools' container using the correct sample test Linux binary as counterpart. The Docker container executable will then return a success/not success value, whereby the script can deduce whether the sample completed successfully or not.

### CODE REVIEW

**Review by pfalcon — DISMISSED** [review]
If the script introduced is supposed to perform heavy-weight operations, like running something in docker, can it please named in way which hints of that, instead of innocent name of "run-sample". If you don't have your own ideas, "run-sample-in-docker" would sound good to me.

File: samples/net/run-sample.sh [hunk 1]
```diff
@@ -0,0 +1,320 @@
+#!/bin/sh
```
Comment (nashif) [inline samples/net/run-sample.sh]: Please put this script under scripts/, this is no place for scripts, for example scripts/net/...

**Review by nashif — CHANGES_REQUESTED** [review]

File: samples/net/run-sample.sh (hunk 1 above)
Comment (pfl) [reply samples/net/run-sample.sh]: Yes, done.

File: samples/net/sockets/echo_client/prj.conf [hunk 2]
```diff
@@ -40,7 +40,7 @@ CONFIG_NET_SHELL=y
 # For linux<->qemu connectivity, create a new conf file and swap the
 # addresses (so that peer address is ending to 2).
 CONFIG_NET_CONFIG_SETTINGS=y
-CONFIG_NET_CONFIG_MY_IPV6_ADDR="2001:db8::2"
-CONFIG_NET_CONFIG_PEER_IPV6_ADDR="2001:db8::1"
-CONFIG_NET_CONFIG_MY_IPV4_ADDR="192.0.2.2"
-CONFIG_NET_CONFIG_PEER_IPV4_ADDR="192.0.2.1"
+CONFIG_NET_CONFIG_MY_IPV6_ADDR="2001:db8::1"
```
Comment (jukkar) [inline samples/net/sockets/echo_client/prj.conf]: Ah, I missed the point when you asked the IP address numbering. I think we need to leave these like this atm so that qemu<->qemu communication works ok.
When compiling the echo-client, we can override the default addresses so can do like this: `cmake .... -DCONFIG_NET_CONFIG_MY_IPV6_ADDR="2001:db8::1" -DCONFIG_NET_CONFIG_PEER_IPV6_ADDR="2001:db8::2" -DCONFIG_NET_CONFIG_MY_IPV4_ADDR="192.0.2.1" -DCONFIG_NET_CONFIG_PEER_IPV4_ADDR="192.0.2.2"`

File: scripts/net/run-sample-tests.sh [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+	fi
+
+	echo "Started Docker network '$network' bridge interface" \
+	     "'$bridge_interface'..."
+    fi
+
+    if ! docker ps | grep "$name" > /dev/null
+    then
+	if ! docker run --hostname=$name --name=$name \
+             --ip=192.0.2.2 --ip6=2001:db8::2 \
```
Comment (jukkar) [inline scripts/net/run-sample-tests.sh]: What about setting the IP addresses somewhere at the beginning of the script, then it is easy to change them if needed.

File: scripts/net/run-sample-tests.sh [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+	    echo "Could not stop Docker network '$network'" >&2
+	fi
+    fi
+}
+
+start_zephyr ()
+{
+    rm -rf build && mkdir build && \
+	cmake -GNinja -DBOARD=native_posix -B build && \
+	ninja -C build
```
Comment (jukkar) [inline scripts/net/run-sample-tests.sh]: We could override the IP addresses here when compiling if needed (related to my earlier comment of echo-client prj.conf)

**Review by jukkar — CHANGES_REQUESTED** [review]
Looks good. I think we need to think more about how to override the default IP numbering scheme.

File: samples/net/sockets/echo_client/prj.conf (hunk 2 above)
Comment (pfl) [reply samples/net/sockets/echo_client/prj.conf]: Ok, let's keep the prj.conf and fix up the addresses assigned to the Docker image instead.

File: scripts/net/run-sample-tests.sh [hunk 5]
```diff
@@ -0,0 +1,262 @@
+#!/bin/sh
```
Comment (nashif) [inline scripts/net/run-sample-tests.sh]: commit message talks about echo_server, this however is not related to echo_server.

File: CODEOWNERS [hunk 6]
```diff
@@ -328,6 +328,7 @@
 /scripts/gen_priv_stacks.py               @andrewboie @agross-oss @ioannisg
 /scripts/gen_syscall_header.py            @andrewboie
 /scripts/gen_syscalls.py                  @andrewboie
+/scripts/net/				  @jukkar @pfl
```
Comment (nashif) [inline CODEOWNERS]: fix that spacing here.

File: scripts/net/run-sample-tests.sh (hunk 3 above)
Comment (pfl) [reply scripts/net/run-sample-tests.sh]: IP addresses (and other arguments) are now set depending on whether Zephyr is a(n echo) server or client.

File: scripts/net/run-sample-tests.sh (hunk 4 above)
Comment (pfl) [reply scripts/net/run-sample-tests.sh]: Yes.

File: scripts/net/run-sample-tests.sh (hunk 5 above)
Comment (pfl) [reply scripts/net/run-sample-tests.sh]: Fixed.

File: CODEOWNERS (hunk 6 above)
Comment (pfl) [reply CODEOWNERS]: Tabs removed, spaces added.

File: samples/net/sockets/echo_client/Kconfig [hunk 7]
```diff
@@ -46,4 +46,11 @@ config NET_SAMPLE_IFACE3_VLAN_TAG
 	  Set VLAN (virtual LAN) tag (id) that is used in the sample
```
Comment (jukkar) [inline samples/net/sockets/echo_client/Kconfig]: The commit "samples: net: echo_client: Add Kconfig option to run a number of times" still says
```
For the echo_client, its prj.conf file defines
the value as 10, in which time the functionality is demonstrated
and the sample can be run in a meaningful way.
```
But as we removed the value from prj.conf this sentence can be removed.

**Review by jukkar — COMMENTED** [review]
Sanitychecker does not like MIT license, should we change this to Apache-2?

File: samples/net/sockets/echo_client/Kconfig (hunk 7 above)
Comment (pfl) [reply samples/net/sockets/echo_client/Kconfig]: Done. And the license.

**Review by jukkar — APPROVED** [review]
LGTM

**Comment by jukkar** [comment]
ping @nashif @pfalcon could you re-review the PR please

**Comment by pfalcon** [comment]
I don't see my comment https://github.com/zephyrproject-rtos/zephyr/pull/19677#pullrequestreview-298668844 being addressed or replied to. However if there's no desire to make things easier to users (just today results of Zephyr UX study were presented at the TSC F2F meeting, and the results aren't stellar), let it be so.

**Comment by pfl** [comment]
> I don't see my comment [#19677 (review)](https://github.com/zephyrproject-rtos/zephyr/pull/19677#pullrequestreview-298668844) being addressed or replied to. However if there's no desire to make things easier to users (just today results of Zephyr UX study were presented at the TSC F2F meeting, and the results aren't stellar), let it be so.

Here I tried to keep the naming generic/balanced. We might figure out something else than a Docker container to run the tests in; if so the script needs a major overhaul, but the "familiar" name would be kept. So I'm inclined to keep the name as not very docker specific as it is now.

**Comment by pfalcon** [comment]
@pfl: Sounds good, thanks for reply.

**Comment by jukkar** [comment]
ping @nashif can you recheck and give +1 if ok.

**Comment by pfl** [comment]
ping @nashif were there any concerns still?

**Comment by nashif** [comment]
few things:
- Can you squash all commits related to the script in one single commit? There is no value in getting all of this into the tree when the script is being submitted the first time. 
- There is no documentation or any guidance on how this script can or should be used.
- when I run it without any options, I am getting very cryptic output:

```
scripts/net/run-sample-tests.sh
$ZEPHYR_BASE /home/anashif/zephyrproject/zephyr
$NET_TOOLS_BASE /home/anashif/zephyrproject/zephyr/../net-tools
'zephyr' not supported
Sample 'zephyr' failed with return value '1'
Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Get http://%2Fvar%2Frun%2Fdocker.sock/v1.40/containers/json: dial unix /var/run/docker.sock: connect: permission denied
Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Get http://%2Fvar%2Frun%2Fdocker.sock/v1.40/networks: dial unix /var/run/docker.sock: connect: permission denied
```

- Don't you need to update west.yml to point to net-tools with docker image?

**Comment by pfl** [comment]
> * Can you squash all commits related to the script in one single commit? There is no value in getting all of this into the tree when the script is being submitted the first time.

Done, there is now one commit for the script itself and another for the echo_client sample.

> * There is no documentation or any guidance on how this script can or should be used.

Added a README to help with that.
 
> * when I run it without any options, I am getting very cryptic output:

Fixed wording of cryptic output, now it mentions that it could not figure out which sample is wanted.

**Comment by pfl** [comment]
ping @nashif ?

**Comment by pfl** [comment]
ping @nashif ?

**Comment by jukkar** [comment]
@nashif can you recheck this one?

File: scripts/net/README.txt [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+As a prerequisite it is assumed that Docker is installed and that the
+'net-tools' Docker container has been created, see the first bullet point
+at the net-tools Docker README file
+https://github.com/zephyrproject-rtos/net-tools/blob/master/README.docker.
+In essence, the following needs to be done:
+
+ * Install Docker
+ * Check out the net-tools project from github
+ * Change working directory to docker/ in the net-tools repository
```
Comment (nashif) [inline scripts/net/README.txt]: is this step needed? I can't run the following command as specified from within the docker/ directory.

File: scripts/net/README.txt [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+ * Run './net-setup.sh --config docker.conf'
```
Comment (nashif) [inline scripts/net/README.txt]: Maybe 'sudo ./net-setup.sh --config docker.conf'

File: scripts/net/README.txt [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+This creates a Docker image called 'net-tools' on your local machine.
```
Comment (nashif) [inline scripts/net/README.txt]: Hmm, nothing happens here.. I get:

```
i9:net-tools((HEAD detached at 354fd05)): sudo ./net-setup.sh --config docker.conf
Using docker.conf configuration file.
Creating zeth
Error response from daemon: network with name net-tools0 already exists
bridge br- does not exist!
br-
```

File: scripts/net/README.txt [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+************
+
+As a prerequisite it is assumed that Docker is installed and that the
+'net-tools' Docker container has been created, see the first bullet point
+at the net-tools Docker README file
+https://github.com/zephyrproject-rtos/net-tools/blob/master/README.docker.
+In essence, the following needs to be done:
+
+ * Install Docker
+ * Check out the net-tools project from github
```
Comment (nashif) [inline scripts/net/README.txt]: net-tools is part of the west manifest, the current sha in west.yml is old and does not have docker/ directory, you need to update west.yml with the latest net-tools and point the user to it, i.e. `../tools/net-tools`. there is no need to checkout a new instance of it.

**Review by nashif — CHANGES_REQUESTED** [review]

**Comment by nashif** [comment]
> -:322: WARNING:TYPO_SPELLING: 'childs' may be misspelled - perhaps 'children'?

maybe you want to address this as well :)

File: scripts/net/README.txt (hunk 9 above)
Comment (jukkar) [reply scripts/net/README.txt]: The script will do sudo if needed.

File: scripts/net/README.txt (hunk 10 above)
Comment (pfl) [reply scripts/net/README.txt]: This is an issue with net-tools, fix is https://github.com/zephyrproject-rtos/net-tools/pull/40

File: scripts/net/README.txt (hunk 8 above)
Comment (pfl) [reply scripts/net/README.txt]: Documentation fixed...

File: scripts/net/README.txt (hunk 11 above)
Comment (pfl) [reply scripts/net/README.txt]: Done.

**Comment by pfl** [comment]
> maybe you want to address this as well :)

Yes, it is horrible with misspelled function names :-)

**Comment by jukkar** [comment]
ping @nashif could you recheck?

**Comment by pfl** [comment]
ping @nashif

**Review by jhedberg — APPROVED** [review]
Looks fine to me. Any chance @nashif that you could take another look before we merge?

**Comment by pfl** [comment]
Ping @nashif ?

**Review by nashif — APPROVED** [review]
